### ISSUE TITLE
[Hardware Requirement] Do I need AMD cpu to contribute to this project?

### ISSUE THREAD

**Issue by coderboyisongithub** [ROCmSoftwarePlatform/hipDNN#81]
I have intel hardware and can I build and work on this project with that?

**Comment by BrianHarrisonAMD** [comment]
I believe you should be able to.
There is a small amount of GPU tests that should be automatically skipped if you don't have a device that works with hip.

You should be able to set up the docker build environment inside the repo, and use that for building / running the tests.
Let us know if you hit issues.

**Comment by coderboyisongithub** [comment]
Well, one more thing. I was expecting tensor implementation and automatic differentiation in this repo as its a neural network library. Its not there. I can contribute to that part.

**Comment by BrianHarrisonAMD** [comment]
Thanks for you interest in this project @coderboyisongithub!

We are in the process of adding some SDK implementations for [Tensor](https://github.com/ROCm/hipDNN/pull/80/files#diff-d048314eb6d4c6dcda8b4bbbc964b0936c9aa02f80e3ba9b5b01f9246c132311) which can be used for testing, or applications (#80 for more details).

Tensor won't need to be part of the frontend API, since we have a schema definition for a tensor, and accept device buffer's as void* through variant packs. If you were looking to extend Tensor in the frontend, then adding some utilities that allow people to easily convert from the usable tensor into TensorAttributes that the frontend works with would be useful!

The Tensors that get passed to the backend, and eventually to the plugins, are intended to describe the structure of a Tensor, but don't provide any implementation. The schema definition is meant to be a simplified description of a tensor, and then consumers are free to leverage the SDK, or convert into a version that makes sense for the plugin (example [here](https://github.com/ROCm/hipDNN/blob/develop/plugins/miopen_legacy_plugin/miopen_tensor.cpp) for the legacy plugin).

Over time, the SDK will be extended to include more reference implementations to use for verification puirposes, and more utilities that can help with the development of plugins.

Also, we will be adding additional operations over time, and that will include backwards directions for things like Batch Normalization, and Convolutions. If you are interested in contributing, we will need to add the schema, frontend implementation (node, attributes, graph implementation), reference implementations for verification, and ideally some plugin implementation to provide a solution for that operation. Adding any piece of that would be useful, but we would likely want to keep any operations that doesn't have implementation out of the graph to start with.

**Comment by coderboyisongithub** [comment]
@BrianHarrisonAMD Thats a sound plan. Also, from your description I speculate you will add a tensor interface/adapter based design at front end of SDK and adapter for various backend tensor implementation. The usual approach I guess? like adapter for Eigen, gpu or the plugins  tensor and etc... And yes the computational graph which will be the 90% of the DNN library. See I can work on that area, I have some insight to what have to be done and can supervise myself. So can you show me something to work on there? Also I would suggest using Xbyak for just in time SIMD generator for SIMD based acceleration

**Comment by BrianHarrisonAMD** [comment]
Yea the expectation is that the frontend will just be describing tensors, and then anyone can contribute general implementations as needed. If it's specific to a plugin implementation, then it should live in the plugin. If it's general, or using an open source framework, then it can be contributed to the SDK / frontend. That being said, we also don't want to increase the dependencies of hipDNN to depend on everything, and some thought will be needed as to how we can add these as header only extensions that are optional. Ideally most tensor implementations would live in the plugins, but I could see us needing to add optional extensions for different tensor implementations to the SDK / frontend overtime as well.

The computational graph portion will depend on plugins I believe.
There will be plugin engines that are doing codegen which I assume would make sense for xbyak.
There will also be plugins that provide static implementations that will only match a specific subset of graphs.
If you were interested in working on a codegen plugin, then you can take a look at the start of the [legacy plugin](https://github.com/ROCm/hipDNN/tree/develop/plugins/miopen_legacy_plugin). 
It's not a codegen plugin, but it has the structure, and set up for engine plugins.

Note: We are still in the early stages of the hipDNN project, and the API hasn't fully solidified yet.
There is the possibility we will need to implement breaking changes, and that could impact any plugins being developed.
